### PR TITLE
feat: implemented the backend for demand and capacity notifications 2.0

### LINKED ISSUE #316: Implement Messages in the backend
**As** ... ,
**I want** ... ,
**so that** ... .

## Hints / Details

- ...

### Outcome
- ...
### Acceptance Criteria
- ...

## Out of Scope

- ...

### PR DESCRIPTION
**Pull request by ReneSchroederLJ**
## Description

- implemented the CRUD endpoints for demand and capacity notications

resolves #316 

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [x] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [x] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

### CODE REVIEW

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/domain/model/Notification.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
+@Setter
+@SuperBuilder
+@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
+@Entity
+@ToString
+public abstract class Notification {
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/domain/model/Notification.java:48]: Please rename the Classes to `DemandAndCapacityNotification` as that's the actual kind of notification. Please reconsider also for Service and Controller. (On controller level maybe the encapsulation makes sense?)

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/service/OwnNotificationService.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    public OwnNotificationService(OwnNotificationRepository ownNotificationRepository, PartnerService partnerService, MaterialPartnerRelationService mpr) {
+        super(ownNotificationRepository, partnerService, mpr);
+    }
+
+    public List<OwnNotification>  findAllByPartnerBpnl(String bpnl) {
+        return repository.findAll().stream().filter(notification -> notification.getPartner().getBpnl().equals(bpnl)).toList();
+    }
+
+    public boolean validate(OwnNotification notification) {
```
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/service/OwnNotificationService.java:42]: Please follow recommendation for readability

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/service/ReportedNotificationService.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            PartnerService partnerService, MaterialPartnerRelationService mpr) {
+        super(reportedNotificationRepository, partnerService, mpr);
+    }
+
+    public List<ReportedNotification> findAllByPartnerBpnl(String bpnl) {
+        return repository.findAll().stream().filter(notification -> notification.getPartner().getBpnl().equals(bpnl))
+                .toList();
+    }
+
+    public boolean validate(ReportedNotification notification) {
```
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/service/ReportedNotificationService.java:44]: Please follow recommendation for readability

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/domain/model/Notification.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+    @ManyToMany()
+    @JoinTable(
+        name = "notification_material",
+        joinColumns = @JoinColumn(name = "notification_uuid"),
+        inverseJoinColumns = @JoinColumn(name = "material_ownMaterialNumber"))
+    @ToString.Exclude
+    protected List<Material> materials;
+
+    protected String text;
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/domain/model/Notification.java:69]: please add pattern check.

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/domain/model/Notification.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    protected UUID uuid;
+
+    protected UUID relatedNotificationId;
+
+    @ManyToOne()
+    @JoinColumn(name = "partner_uuid")
+    @ToString.Exclude
+    @NotNull
+    protected Partner partner;
+
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/domain/model/Notification.java:60]: missing property: `sourceNotificationId` 
Is uuid = notificationId?

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/dto/NotificationDto.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    private UUID uuid;
+
+    private UUID relatedNotificationId;
+
+    @Pattern(regexp = PatternStore.BPNL_STRING)
+    private String partnerBpnl;
+
+    private List<@Pattern(regexp = PatternStore.NON_EMPTY_NON_VERTICAL_WHITESPACE_STRING) String> affectedMaterialNumbers;
+
+    private String text;
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/dto/NotificationDto.java:53]: pattern validation needed

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/dto/NotificationDto.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import lombok.ToString;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@ToString
+public class NotificationDto implements Serializable {
+    private UUID uuid;
+
+    private UUID relatedNotificationId;
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/dto/NotificationDto.java:46]: `sourceNotificationId` missing

**Review by tom-rm-meyer-ISST — CHANGES_REQUESTED** [review]
Overall great work. Had a few findings. 

Sidenote: Please always mark the dependency and license header things in the pr on your own prior to requesting the review :)

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/domain/model/Notification.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ReneSchroederLJ) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/domain/model/Notification.java]: Yes. The interface for handling communication with the EDC is too extensive to include in the same PR and will be handled separately

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/dto/NotificationDto.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@ToString
+public class NotificationDto implements Serializable {
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/notification/logic/dto/NotificationDto.java:43]: @ReneSchroederLJ is not renaming this Dto similar to services, controllers, model on purpose? Else I propose to also rename and maybe also change the route of the controller to dcm-notification to use the same wording everywhere.

**Review by tom-rm-meyer-ISST — COMMENTED** [review]
Please see new comment. Regarding bpnl sanitizing: not sure if you missed the comment, please check. If you propose to not go with sanitizing the controllers parameter, please elaborate why.

**Comment by ReneSchroederLJ** [comment]
> Please see new comment. Regarding bpnl sanitizing: not sure if you missed the comment, please check. If you propose to not go with sanitizing the controllers parameter, please elaborate why.

I'm unsure what you mean regarding the BPNL sanitizing. Is the pattern check I added to the parameter not enough?

**Comment by tom-rm-meyer-ISST** [comment]
> > Please see new comment. Regarding bpnl sanitizing: not sure if you missed the comment, please check. If you propose to not go with sanitizing the controllers parameter, please elaborate why.
> 
> I'm unsure what you mean regarding the BPNL sanitizing. Is the pattern check I added to the parameter not enough?

Ah sorry, you're right. Got confused due to the outdated conversation.

**Review by tom-rm-meyer-ISST — APPROVED** [review]
LGTM, great contribution! :)

**Pull request merged**